x/distribution/keeper/hooks.go

```diff
@@ -1,146 +1,60 @@
-			panic(err)
-		}
-	}
-}
-
-// Withdraw all validator rewards
-func (k Keeper) AfterValidatorBonded(ctx sdk.Context, valAddr sdk.ValAddress) {
-	lastPower := k.stakeKeeper.GetLastValidatorPower(ctx, valAddr)
-	if !lastPower.Equal(sdk.ZeroInt()) {
-		panic("expected last power to be 0 for validator entering bonded state")
-	}
-	k.BeforeValidatorModified(ctx, valAddr)
-}
-
-// Sanity check, very useful!
-func (k Keeper) AfterValidatorPowerDidChange(ctx sdk.Context, valAddr sdk.ValAddress) {
-	vi := k.GetValidatorDistInfo(ctx, valAddr)
-	if vi.FeePoolWithdrawalHeight != ctx.BlockHeight() {
-		panic(fmt.Sprintf("expected validator (%v) dist info FeePoolWithdrawalHeight to be updated to %v, but was %v.",
-			valAddr.String(), ctx.BlockHeight(), vi.FeePoolWithdrawalHeight))
-	}
-}
-
-// Withdrawal all validator distribution rewards and cleanup the distribution record
-func (k Keeper) AfterValidatorRemoved(ctx sdk.Context, valAddr sdk.ValAddress) {
-	k.RemoveValidatorDistInfo(ctx, valAddr)
-}
-
-//_________________________________________________________________________________________
-
-// Create a new delegator distribution record
-func (k Keeper) BeforeDelegationCreated(ctx sdk.Context, delAddr sdk.AccAddress,
-	valAddr sdk.ValAddress) {
-
-	ddi := types.DelegationDistInfo{
-		DelegatorAddr:           delAddr,
-		ValOperatorAddr:         valAddr,
-		DelPoolWithdrawalHeight: ctx.BlockHeight(),
-	}
-	k.SetDelegationDistInfo(ctx, ddi)
-}
-
-// Withdrawal all validator rewards
-func (k Keeper) BeforeDelegationSharesModified(ctx sdk.Context, delAddr sdk.AccAddress,
-	valAddr sdk.ValAddress) {
-
-	if err := k.WithdrawDelegationReward(ctx, delAddr, valAddr); err != nil {
-		panic(err)
-	}
-}
-
-// Withdrawal all validator distribution rewards and cleanup the distribution record
-func (k Keeper) BeforeDelegationRemoved(ctx sdk.Context, delAddr sdk.AccAddress,
-	valAddr sdk.ValAddress) {
-	// Withdraw validator commission when validator self-bond is removed.
-	// Because we maintain the invariant that all delegations must be removed
-	// before a validator is deleted, this ensures that commission will be withdrawn
-	// before the validator is deleted (and the corresponding ValidatorDistInfo removed).
-	// If we change other parts of the code such that a self-delegation might remain after
-	// a validator is deleted, this logic will no longer be safe.
-	// TODO: Consider instead implementing this in a "BeforeValidatorRemoved" hook.
-	if valAddr.Equals(sdk.ValAddress(delAddr)) {
-		feePool, commission := k.withdrawValidatorCommission(ctx, valAddr)
-		k.WithdrawToDelegator(ctx, feePool, delAddr, commission)
-	}
-
-	k.RemoveDelegationDistInfo(ctx, delAddr, valAddr)
-}
-
-//_________________________________________________________________________________________
-
 // Wrapper struct
 type Hooks struct {
 	k Keeper
 }
 
 var _ sdk.StakingHooks = Hooks{}
 
-// New Validator Hooks
+// Create new distribution hooks
```

Comment:
Do we still need all these hooks? I thought one of the benefits to F1 was that we could remove some of this code?

Reply:
After reading a bit more closely it looks like we have. We need to revisit the hooks implementation and interfaces at some point.

Reply:
We did - in particular, note that changes to validators or delegations only affect the validator or the delegation themselves, never other validators or delegations.

F1 can't get rid of all the hooks though - that will be true of any non-iterating fee distribution I think.

Reply:
Agreed, fundamentally a fee distribution and staking have to do things on this "withdraw" action, and on this "slash" action.

What would be preferrable imo is instead of hooks where its one module in charge, we instead have like "event registration", and multiple modules subscribe to the event. Then things like column hashing our values cross-modules becomes trivial, and that is a massive state saving. In particular it lets fee distribution not create an extra state element for a single int, per delegation.